### PR TITLE
Added X for unknown to aa.py and tests, allow testing for upper case.

### PR DESCRIPTION
**Pull request by bamueh**

### CODE REVIEW

File: test/test_fasta.py [hunk 1]
```diff
@@ -394,11 +394,42 @@ def testOnlyCheckSomeAlphabets(self):
...
+        """
+        data = '\n'.join(['>id1', 'actg'])
+        mockOpener = mockOpen(read_data=data)
+        with patch.object(builtins, 'open', mockOpener):
+            reads = list(FastaReads('filename.fasta', upperCase=True))
+            self.assertEqual([AARead('id1', 'ACTG')], reads)
+
+    def testDontConvertLowerToUpperCaseIfNotSpecified(self):
+        """
+        A read needs to be converted from lower to upper case if specified.
```
Comment (terrycojones) [inline test/test_fasta.py]: Copy / paste error.

File: dark/fasta.py [hunk 2]
```diff
@@ -85,11 +85,15 @@ class FastaReads(Reads):
     @param checkAlphabet: An C{int} or C{None}. If C{None}, alphabet checking
         will be done on all reads. If an C{int}, only that many reads will be
         checked. (Pass zero to have no checks done.)
+    @param upperCase: If not C{None}, the reads will be converted to upper
+        case.
     """
-    def __init__(self, _file, readClass=DNARead, checkAlphabet=None):
+    def __init__(self, _file, readClass=DNARead, checkAlphabet=None,
+                 upperCase=None):
```
Comment (terrycojones) [inline dark/fasta.py]: The default value should be `False`, as it's a boolean. It doesn't matter really, but it makes things a little clearer for the reader.

File: dark/aa.py [hunk 3]
```diff
@@ -87,6 +89,7 @@
 NEGATIVE = 0x0400
 POLAR = 0x0800
 NONE = 0x1000
+UNKNOWN = 0x0000
```
Comment (terrycojones) [inline dark/aa.py]: You've added this, but is it used anywhere?

**Comment by terrycojones** [comment]
I don't know what you're trying to achieve in this ticket. There are two things going on - the upper casing and the addition of `X` to the AAs.  What is/are the problem/s you're trying to solve?

I don't like the addition of `X` to the AAs. It's not an AA. It feels like it might be better to elsewhere deal with sequences that have `X` in them (e.g., in some of my code I use `try/except` to deal with not being able to look up AA properties). Or maybe we should have a class like `AAWithXRead` that simply subclasses `AARead` but adds an `X` to its `self.ALPHABET`? But maybe it's just cleaner/simpler to do it the way you've done it.  It somehow feels wrong to me.

If we leave out `X` (as it is now) then anyone who has one has to deal with things like `KeyError` but OTOH if we put it in (as in your branch), anyone who wants to do things with AAs has to deal with a non-AA in the mix.

Anyway, if I knew what problem you were trying to solve here it might be easier to see what the best approach is.

**Comment by bamueh** [comment]
A dataset I was using had an X in it (for unknown amino acid), so I put it in. I think it's useful not having to do with the key error.
The upper case checking is in there because one of Christians dataset is in lower case, and because the  finders are case-sensitive, they didn't work, and I don't want us to accidentally have an error / not find any hits because of that.

**Comment by terrycojones** [comment]
The upper case thing makes sense, is localized to just the reading of FASTA files, and is easy to turn on / off.

The X change makes me uneasy. It's in there at the lowest level, and requires weird things to be put in there (like all the NONE properties) seeing as it is not an AA. On reading nucleotide sequences we don't allow an X or any of the other many (see http://www.boekhoff.info/?pid=data&dat=fasta-codes) letters that indicate uncertainty. It feels wrong. Anyway, I guess you could merge it and we can live with it.

It's not much work to deal with the key error, you can just use `get`. E.g.

```python
from dark.aa import PROPERTIES

for x in sequence:
    properties = PROPERTIES.get(x)
    if properties:
        # Etc
```

That's just one more line of code.

**Comment by bamueh** [comment]
So are you uneasy with x being part of the properties or with it being part of the alphabet?

**Comment by terrycojones** [comment]
Actually, the code I wrote above (using `get`) is probably simpler than you'd write it if `X` were not added to the AAs. With your change, you'd end up writing:

```python
from dark.aa import PROPERTIES, NONE

for x in sequence:
    properties = PROPERTIES.get(x)
    if properties is not NONE:
        # Etc
```

So one way or another the unusual value (either `X` in the current code or `NONE` in the new code) ends up having to be explicitly checked for.

And, you've added to the `CODONS` dict, putting in more `X`s instead of nucleotides. Next you'll be putting `X` into our nucleotide alphabets... It's a creeping mess :-)

**Comment by bamueh** [comment]
I don't understand where you want to make the change above?

**Comment by terrycojones** [comment]
> So are you uneasy with x being part of the properties or with it being part of the alphabet?

I don't think it should be in the `PROPERTIES`, `CODONS`, or `PROPERTY_DETAILS` dicts, or in a key value in `CODONS`. Plus I don't see how the change makes using the `aa.py` code any easier (see above).

It's not perfect, but adding a read class that is a subclass of AARead (with an ALPHABET that has the AAs plus `X`) would make it simple to read a FASTA file of AA seqs that contains `X`. We could provide that ourselves but the user could also provide it, as it's just a couple of lines:

```python
from dark.reads import AARead
from dark.aa import AA_LETTERS
from dark.fasta import FastaReads

class ReadWithXs(AARead):
    ALPHABET = set(AA_LETTERS + ['X'])

for read in FastaReads(filename, readClass=ReadWithXs):
    # Etc.
```

That allows people to easily/flexibly add `X` as well as `-` or anything else that might be in their FASTA.

**Comment by terrycojones** [comment]
> I don't understand where you want to make the change above?

That was just code to illustrate how our code would be used.  In the existing case, you need to check for missing keys. In the proposed change you need to check that the value you get back (e.g., from `PROPERTIES`) is not some other special value (`NONE`). I don't see any win there, only the negative from building `X` in at a low level when it simply isn't an AA.

**Comment by bamueh** [comment]
OK, I understand. But if you do that, you'll still get an error if a landmark finder tries to use the properties.

**Comment by terrycojones** [comment]
> But if you do that, you'll still get an error if a landmark finder tries to use the properties.

What's an example?  In `alpha_helix.py` it does this:

```python
        result = []
        append = result.append
        get = PROPERTIES.get

        for residue in sequence:
            # 'O' comes from hydrophObic, 'I' from hydrophIlic
            append('O' if (get(residue, 0x0) & HYDROPHOBIC) else 'I')
        return ''.join(result)
```

So it's dealing with an `X` (or a `-` or anything else unrecognized) via `get`.

**Comment by bamueh** [comment]
Ah, ok.

File: test/test_aa.py [hunk 4]
```diff
@@ -190,11 +190,11 @@ def testCodonLength(self):
 
     def testCodonContent(self):
         """
-        Codons must only contain the letters A, T, G, C.
+        Codons must only contain the letters A, T, G, C and X for unknown.
```
Comment (terrycojones) [inline test/test_aa.py]: The two mentions of `X` should go away here.

**Pull request merged**